### PR TITLE
[DO NOT MERGE, JUST FOR REVIEW] Done from Simon Guo

### PR DESCRIPTION
**Pull request by SimonKwok**
Made small change on unit test as well, please help to review and let me know if there is anything incorrect. Thanks for preparing this.

Simon Guo

### CODE REVIEW

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 1]
```diff
@@ -27,7 +31,10 @@ public DataProcessor(List<RoadAccident> roadAccidentList){
      * @return
      */
     public RoadAccident getAccidentByIndex7(String index){
-        return null;
+        Optional<RoadAccident> foundRoadAccident = roadAccidentList.stream()
+                .filter(accident -> accident.getAccidentId().equals(index))
+                .findFirst();
+        return foundRoadAccident.isPresent()?foundRoadAccident.get():null;
```
Comment (bug-hunterx) [inline src/main/java/com/epam/processor/DataProcessor.java]: That's one way of getting values from Optional. Another one is to use https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElse-T- 

optional.orElse(null);

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 2]
```diff
@@ -1,11 +1,15 @@
 package com.epam.processor;
```
Comment (bug-hunterx) [inline src/main/java/com/epam/processor/DataProcessor.java:1]: General point for entire file - all methods with 7 should be implemented with Java 7 api, ie without streaming. in this way you will see what's the difference is and will be able to decide by yourself in future what to choose.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 3]
```diff
@@ -40,7 +47,10 @@ public RoadAccident getAccidentByIndex7(String index){
      * @return
      */
     public Collection<RoadAccident> getAccidentsByLocation7(float minLongitude, float maxLongitude, float minLatitude, float maxLatitude){
-        return null;
+        return roadAccidentList.stream()
+                .filter(accident -> isInScope(accident.getLongitude(), minLongitude, maxLongitude)
```
Comment (bug-hunterx) [inline src/main/java/com/epam/processor/DataProcessor.java]: isInScope - very good move :+1:

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 4]
```diff
@@ -51,7 +61,8 @@ public RoadAccident getAccidentByIndex7(String index){
      * @return
      */
     public Map<String, Long> getCountByRoadSurfaceCondition7(){
-        return null;
+        return roadAccidentList.stream()
+                .collect(Collectors.groupingBy(RoadAccident::getRoadSurfaceConditions, Collectors.counting()));
```
Comment (bug-hunterx) [inline src/main/java/com/epam/processor/DataProcessor.java]: so you are familiar with java 8? :smile:  
I think I did not cover static methods invocations ...

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 5]
```diff
@@ -71,7 +87,10 @@ public RoadAccident getAccidentByIndex7(String index){
      * @return
      */
     public Multimap<String, String> getAccidentIdsGroupedByAuthority7(){
-        return null;
+        Multimap<String, String> resultMap = ArrayListMultimap.create();
+        roadAccidentList.stream()
+                .forEach(accident -> resultMap.put(accident.getDistrictAuthority(), accident.getAccidentId()));
+        return resultMap;
```
Comment (bug-hunterx) [inline src/main/java/com/epam/processor/DataProcessor.java:116]: I actually wanted you to use java 7 with Guava, but this is really nice combination of streaming and guava multimap. Good job

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 6]
```diff
@@ -93,31 +115,46 @@ public RoadAccident getAccidentByIndex(String index){
...
 
     /**
      * To match streaming operations result, return type is a java collection instead of multimap
      * @return
      */
     public Map<String, List<String>> getAccidentIdsGroupedByAuthority(){
-        return null;
+        return roadAccidentList.stream()
+                .map(roadAccident -> roadAccident.getDistrictAuthority())
+                .collect(Collectors.groupingBy(String::toString, Collectors.toList()));
```
Comment (bug-hunterx) [inline src/main/java/com/epam/processor/DataProcessor.java]: Yea - you passed my tests, but this is not doing what's supposed to. At collect time you don't have information about accidentId anymore, so think about what your Collectors.toList is actually collecting? :smirk:

File: src/main/java/com/epam/processor/DataProcessor.java (hunk 4 above)
Comment (SimonKwok) [inline src/main/java/com/epam/processor/DataProcessor.java]: Actually not. I was wondering what is needed here, so I click into the Collectors.groupingBy. And I found some example there.

File: src/main/java/com/epam/processor/DataProcessor.java (hunk 1 above)
Comment (SimonKwok) [inline src/main/java/com/epam/processor/DataProcessor.java]: cool, more clean.

File: src/main/java/com/epam/processor/DataProcessor.java (hunk 2 above)
Comment (SimonKwok) [inline src/main/java/com/epam/processor/DataProcessor.java:1]: Oh, I will correct it later.

File: src/main/java/com/epam/processor/DataProcessor.java (hunk 6 above)
Comment (SimonKwok) [inline src/main/java/com/epam/processor/DataProcessor.java]: haha, sorry for this one. actually, this one took me more than 30 mins and still can't work it out. Let me try again later.

File: src/main/java/com/epam/processor/DataProcessor.java (hunk 2 above)
Comment (SimonKwok) [inline src/main/java/com/epam/processor/DataProcessor.java:1]: Done.

File: src/main/java/com/epam/processor/DataProcessor.java [hunk 7]
```diff
@@ -93,31 +138,48 @@ public RoadAccident getAccidentByIndex(String index){
...
      * To match streaming operations result, return type is a java collection instead of multimap
      * @return
      */
     public Map<String, List<String>> getAccidentIdsGroupedByAuthority(){
-        return null;
+        Map<String, List<String>> resultMap = new HashMap<String, List<String>>();
+        Multimap<String,String> authorityToIdsMap = ArrayListMultimap.create();
+        roadAccidentList.stream().forEach(roadAccident -> authorityToIdsMap.put(roadAccident.getDistrictAuthority(), roadAccident.getAccidentId()));
+        authorityToIdsMap.keySet().stream().forEach(authority -> resultMap.put(authority, ImmutableList.copyOf(authorityToIdsMap.get(authority))));
+        return resultMap;
```
Comment (bug-hunterx) [inline src/main/java/com/epam/processor/DataProcessor.java:182]: You can do better :smirk: 
this task can be done in one line - like almost all streaming tasks

Hint: check Collectors.groupingBy and Collectors.mapping apis

So ideally I want to see one line
return accidentsList.stream()...... ;

Can you do it? :smile:

Comment (SimonKwok) [inline src/main/java/com/epam/processor/DataProcessor.java:182]: Sorry, I tried but can't. I saw alexeymatyas's PR so I know how to do it now. It is -
mapping(RoadAccident::getAccidentId, toList())
BUT, TBH, I don't like the API of 'Collectors.mapping'. Actually, I was seeking some solution like - toList(map(accident-> accident.getAccidentId()))

File: src/main/java/com/epam/processor/DataProcessor.java (hunk 1 above)
Comment (gavin2lee) [inline src/main/java/com/epam/processor/DataProcessor.java]: Hi,how could u use 8's api here?? Optional was introduced in Java 8, not 7. :)

File: src/main/java/com/epam/processor/DataProcessor.java (hunk 7 above)
Comment (bug-hunterx) [inline src/main/java/com/epam/processor/DataProcessor.java:182]: I see your point - mapping collectors is quite confusing.  

You can always write your own collector :) That will be a good exercise to better understand how they work.

**Comment by bug-hunterx** [comment]
Closing as homework is accepted
